adapters/repos/db/vector/hnsw/packedconn/connections.go

```diff
@@ -0,0 +1,327 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package packedconn
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"sort"
+)
+
+// Data order
+// | Start | Len        | Description |
+// | ----- | ---------- | ------------|
+// | 0     | 1          | uint8 indicating count of layers of object |
+// | 1     | 3 * layers | uint8 len indicator, followed by uint16 offset indicator |
+// | dyn.  | dyn        | all layers, starting with highest layer first
+
+type Connections struct {
+	data []byte
+	buff []byte
+}
+
+const (
+	layerPos    = 0
+	initialSize = 50
+)
+
+func NewWithMaxLayer(maxLayer uint8) (*Connections, error) {
+	if int(maxLayer)+1 > math.MaxUint8 {
+		return nil, fmt.Errorf("max supported layer is %d",
+			math.MaxUint8-1)
+	}
+	c := Connections{
+		// TODO: low initial size and grow dynamically
+		data: make([]byte, initialSize+1+(maxLayer+1)*3),
+	}
+
+	c.initLayers(maxLayer)
+
+	c.buff = make([]byte, 16)
+
+	return &c, nil
+}
+
+func (c *Connections) AddLayer() {
```

Comment:
`AddLayer()` is called multiple times in a loop if multiple layers have to be added at once. Maybe it is worth to add an argument to the method (number of layers to be added) to perform all those expanding, shifting, etc at once.